pyro/distributions/transforms/householder.py

```diff
@@ -218,14 +224,18 @@ def __init__(self, input_dim, nn, count_transforms=1):
         self.count_transforms = count_transforms
 
     def condition(self, context):
-        # u_unnormed ~ (count_transforms, input_dim)
-        # Hence, input_dim must divide
-        u_unnormed = self.nn(context)
-        if self.count_transforms == 1:
-            u_unnormed = u_unnormed.unsqueeze(-2)
-        else:
-            u_unnormed = torch.stack(u_unnormed, dim=-2)
-        return ConditionedHouseholder(u_unnormed)
+        def params():
+            # u_unnormed ~ (count_transforms, input_dim)
+            # Hence, input_dim must divide
+            u_unnormed = self.nn(context)
+            if self.count_transforms == 1:
```

Comment:
Not part of this PR, but it is an antipattern to make return type depend on a scalar value. Doing so breaks the ability to write generic code using this interface. Has this behavior been released yet?

Reply:
Yes I'm afraid it's been released... Can you elaborate on why this would break the ability to write generic code? I think the underlying problem is actually how `DenseNN` shapes it's outputs, and this is something we should think about more carefully during a major refactoring

Reply:
**warning:** the following is highly subjective

An example of some generic code you might like to write is, say, writing a `ConditionalCatTransform` that concatenates two conditional transforms. You might want to fuse their neural nets to save on tensor ops and share strength. How would you do this? Maybe
```py
class ConditionalCatTransform(...):
    def __init__(self, parts):
        super().__init__(self, cache_size=1)
        self.parts = parts  # ignoring ModuleList magic for demo purposes
        self.nn = memoize(cat_dense_nn([part.nn for part in self.parts]))
        end = 0
        for part in self.parts:
            beg, end = end, end + part.count_transforms
            # The following line assumes DenseNN returns a tuple:
            part.nn = lambda context: self.nn(context)[beg:end]
```
then we could define helpers `memoize = functools.lru_cache(max_size=1)` and
```py
def cat_dense_nn(parts):
    input_dims = parts[0].input_dims
    hidden_dims = parts[0].hidden_dims
    param_dims = sum([part.param_dims for part in parts], [])
    return DenseNN(input_dims, hidden_dims, param_dims) 
```
That's what we could have written if the types were consistent. It would have been pretty simple generic code. But it looks like `DenseNN` returns an output type that depends on an int value, so we would need to complicate our wrapper with extra logic:
```diff
  class ConditionalCatTransform(...):
      def __init__(self, parts):
          ...
          for part in self.parts:
              beg, end = end, end + part.count_transforms
-             part.nn = lambda context: self.nn(context)[beg:end]
+             if beg + 1 == end:
+                 part.nn = lamba context: self.nn(context)[beg]
+             else:
+                 part.nn = lambda context: self.nn(context)[beg:end]
```
Now that's just a little more complex. But I feel, as an author of complex code, that the complexity tax is best paid by one-off code, so that abstractions can be built tax free. I think this tax structure is one of the main differences between programming languages, e.g. R and MATLAB tend to tax library code, whereas Python and C tend to tax one-off code.

Reply:
Okay, let's think about this when we have a chance to overhaul `DenseNN`